[Enhancement] Add CollectionView progress behind feature flag

hartez opened the pull request:
### Description of Change ###

Adds progress on #3172 behind the CollectionView_Experimental flag.

### Issues Resolved ### 

- #3172, partially.

### API Changes ###

See #3172, though not all of those changes are complete yet.

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android
- UWP

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### PR Checklist ###

- [x] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

adrianknight89 commented on Xamarin.Forms.ControlGallery.Android/FormsAppCompatActivity.cs:
```diff
@@ -43,7 +43,9 @@ protected override void OnCreate(Bundle bundle)
-			Forms.SetFlags("Fake_Flag"); // So we can test for flag initialization issues
+			// Fake_Flag is here so we can test for flag initialization issues
+			// CollectionView lets us test CollectionView stuff until it's officially released
+			Forms.SetFlags("Fake_Flag", "CollectionView_Experimental"); 
```
Why are these two grouped together as opposed to being separate statements where the latter part goes under `#if TEST_EXPERIMENTAL_RENDERERS`?

hartez replied:
Good point; I've added CollectionView to the "experimental" list.

adrianknight89 replied:
Would `CollectionView_Experimental` still be part of `Forms.SetFlags("Fake_Flag", "CollectionView_Experimental");`?

hartez replied:
Yes.  `TEST_EXPERIMENTAL_RENDERERS` is just there to allow our CI to build and UI test a version of Control Gallery with the fast renderers (a better name for it would have been `TEST_FAST_RENDERERS`).  CollectionView should be available for testing in both versions of Control Gallery.

adrianknight89 commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,104 @@
+
+		// TODO hartez 2018/05/31 15:56:23 Should these just be called Vertical and Horizontal (without List)?	
+		public static readonly IItemsLayout VerticalList = new ListItemsLayout(ItemsLayoutOrientation.Vertical); 
+		public static readonly IItemsLayout HorizontalList = new ListItemsLayout(ItemsLayoutOrientation.Horizontal);
```
Should these be lazy loaded?

hartez replied:
I doubt these constructors are more expensive than a `Lazy<T>`.

adrianknight89 commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,104 @@
+		{
+			var orientation = Orientation == ItemsLayoutOrientation.Horizontal ? "Horizontal Grid" : "Vertical Grid";
+			
+			return $"{orientation}, {nameof(Span)} {Span}";
```
Should we put a `:` after `{nameof(Span)}`?

adrianknight89 commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,134 @@
+		}
+
+		public static readonly BindableProperty EmptyViewTemplateProperty =
+			BindableProperty.Create(nameof(EmptyView), typeof(DataTemplate), typeof(ItemsView), null);
```
`nameof` should use `EmptyViewTemplate`

hartez replied:
Good catch! Fixed.

adrianknight89 commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,134 @@
+		// TODO hartez 2018/06/24 11:37:00 Give DisplayMemberPath some thought	
+
+		public static readonly BindableProperty EmptyViewProperty =
+			BindableProperty.Create(nameof(EmptyView), typeof(object), typeof(ItemsView), null);
```
Why is this an object instead of `View`?

hartez replied:
Because EmptyView can be a View (when EmptyViewTemplate is not set), or it can be the binding context (when EmptyViewTemplate _is_ set).

adrianknight89 commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,134 @@
+		// ((because setting Carousel to grid is ... weird? And by default it just won't do anything.))
+		// And allow CollectionView to use a broader set of Layout options
+		// So the Bindable property only exists at the CarouselView/CollectionView (i.e., concrete class) level
+		// but some version of IItemsLayout is still here?
```
This could also be enforced through exceptions to restrict usage, but I'd use it as a last resort.

adrianknight89 commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,134 @@
+			var minimumSize = new Size(40, 40);
+
+			var maxWidth = Math.Min(Device.Info.ScaledScreenSize.Width, widthConstraint);
+			var maxHeight = Math.Min(Device.Info.ScaledScreenSize.Height, heightConstraint);
```
Somewhat confused by the naming of these variables. Should they be named `minWidth` and `minHeight` instead or `maxAllowedWidth` and `maxAllowedHeight`?

hartez replied:
I don't see why they are confusing. The maximum possible width of the control is the width of the screen, unless a smaller width constraint is given. Same for the height.

adrianknight89 replied:
The confusing part for me was seeing `max` in variable names and then seeing `Math.Min` in the initialization. In a totally unrelated example, we could do `var maxWidth = Math.Max(Device.Info.ScaledScreenSize.Width, widthConstraint);` and the variable name would still make sense. Not a huge deal though, esp. after seeing the init logic. :)

adrianknight89 commented on Xamarin.Forms.Core/Items/ScrollToRequestEventArgs.cs:
```diff
@@ -0,0 +1,40 @@
+		public ScrollToMode Mode { get; }
+
+		public ScrollToPosition ScrollToPosition { get; }
+		public bool Animate { get; }
```
Verbs are best for method names. I'd prefer `Animated` or `IsAnimated`.

hartez replied:
Agreed. Renaming to `IsAnimated`.

adrianknight89 commented on Xamarin.Forms.Platform.iOS/CollectionView/TemplatedCell.cs:
```diff
@@ -0,0 +1,73 @@
+
+		public IVisualElementRenderer VisualElementRenderer { get; private set; }
+
+		public override void Constrain(nfloat constant)
```
Maybe needs to be renamed to `ConstrainTo` like an earlier example.

adrianknight89 commented on Xamarin.Forms.Platform.iOS/CollectionView/ListSource.cs:
```diff
@@ -0,0 +1,25 @@
+		{
+			foreach (object item in enumerable)
+			{
+				Add(item);
```
No way to leverage `AddRange`?

hartez replied:
Your comment made me curious about the performance of AddRange/InsertRange here; using either one would require doing something like `.Cast<object>` from LINQ, but the internal storage growth might be more efficient. So I whipped up a quick test: https://github.com/hartez/ListSourceTest

From my quick-and-dirty testing, adding the items one-by-one in a `foreach` took about 3/4 the time of AddRange/InsertRange. Feel free to clone that repo and look at my testing; I might have missed something obvious. But it looks like this is the fastest option (which surprises me).

Disclaimer: testing was done on a desktop using StopWatch, so this is pseudo-scientific at best.

adrianknight89 replied:
I ran your project, and I'm getting similar results. Looks like casting `IEnumerable` to `IEnumerable<T>` via LINQ is costly. 

I tried replacing the `foreach` loop with a `MoveNext()/Current` logic, but there was no performance benefit. Also, your last sample is the most efficient partly because the other two run beforehand and the underlying data source does not change. If you comment them out and leave the last one, it will take much longer to finish. There is some initialization overhead which gets gobbled up by the other tests.

Having said that, I think `foreach` here is the right choice.

samhouts commented:
Not a blocker for this phase, but it does crash on iOS9.3
![image](https://user-images.githubusercontent.com/7827070/47825331-8f776200-dd2e-11e8-8dfe-45a5f051be87.png)

hartez commented:
> Not a blocker for this phase, but it does crash on iOS9.3 (clicked the first button)
> ![image](https://user-images.githubusercontent.com/7827070/47825331-8f776200-dd2e-11e8-8dfe-45a5f051be87.png)

Fixed with https://github.com/xamarin/Xamarin.Forms/pull/4265/commits/16d2b199817ce4bb82294c16ceedf9152a385b5f

StephaneDelcroix commented on Xamarin.Forms.Core/Application.cs:
```diff
@@ -206,6 +219,19 @@ public ResourceDictionary Resources
 
 		public event EventHandler<Page> PageDisappearing;
 
+		public static void SetFlags(params string[] flags)
```
we already have the flags in Core at the Device level. Application, or Device: pick one

hartez replied:
D'oh. Somehow I missed that we already had them in Device.

hartez replied:
Yeah, that was dumb. Fixed.

StephaneDelcroix commented on Xamarin.Forms.Core/Flags.cs:
```diff
@@ -0,0 +1,39 @@
+{
+	public static class CoreFlags
+	{
+		internal const string CollectionViewExperimental = "CollectionView_Experimental";
```
so, the magic string is internal, but the helper is public ?

I'd drop this class, and move those into the CollectionView class. Keep the string internal, or we won't be able to drop it in the future, and turn the helper internal or at least not EditorBrowseable

hartez replied:
Done.

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,103 @@
+namespace Xamarin.Forms
+{
+	// TODO hartez 2018/05/31 11:49:49 Move the stuff in ItemsControl.cs to the appropriate files
+	public interface IItemsLayout : INotifyPropertyChanged {}
```
![image](https://user-images.githubusercontent.com/313003/48021273-6b55bf80-e138-11e8-9cb0-4db84e762353.png)

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,103 @@
+
+	public class ListItemsLayout : ItemsLayout
+	{
+		// TODO hartez 2018/08/29 17:28:42 Consider changing this name to LinearItemsLayout; not everything using it is a list (e.g., Carousel)	
```
👍

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,103 @@
+		{
+		}
+
+		// TODO hartez 2018/05/31 15:56:23 Should these just be called Vertical and Horizontal (without List)?	
```
👍

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,103 @@
+
+		// TODO hartez 2018/08/29 20:31:54 Need something like these previous two, but as a carousel default	
+
+		public override string ToString()
```
this was probably there for debugging purposes. it's time to get rid of it (if we ship strings, we need to consider internationalization)

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,103 @@
+			Span = span;
+		}
+
+		public override string ToString()
```
drop

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,103 @@
+			set => SetValue(SpanProperty, value);
+		}
+
+		public GridItemsLayout([Parameter("Span")] int span, [Parameter("Orientation")] ItemsLayoutOrientation orientation) :
```
why are you doing the `[Parameter]` thing ?

hartez replied:
Leftover from when Span wasn't mutable. Fixed now.

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsControl.cs:
```diff
@@ -0,0 +1,103 @@
+
+	public class GridItemsLayout : ItemsLayout
+	{
+		public static readonly BindableProperty SpanProperty =
```
shouldn't Span be an attached BP? or am I missing something ?

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,142 @@
+	{
+		public static bool Is(this PropertyChangedEventArgs args, BindableProperty property)
+		{
+			CoreFlags.VerifyCollectionViewFlagEnabled();
```
you can't have both this as an extension method AND contains CV specific check  ¯\_(ツ)_/¯

hartez replied:
Why not? What if I make it not EditorBrowseable?

StephaneDelcroix replied:
your class is `PropertyChangedEventArgsExtension`. given the name and the API I might want to use it for any property changed. but it won't, unless the (unrelated to the class) flag is set. 👎 

also, writing `if (args.Is(Label.TextProperty))` is a bit odd, due to the already existing meaning of `is`. So I read this as `if (args is Label.TextProperty)`...

hartez replied:
Moved this class to an internal helper on each platform, so not accidentally discoverable and no need for the flag.

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,142 @@
+namespace Xamarin.Forms
+{
+	// TODO hartez 2018/06/23 13:42:22 Trying this out for a nicer read in OnElementPropertyChanged, not sure if I like it yet	
+	public static class PropertyChangedEventArgsExtensions
```
change this to internal while we are unsure

hartez replied:
I can't, the renderers need visibility to it.

StephaneDelcroix replied:
If this is a helper specific to renderers but that we don't want to expose publicly, we can copy the helper class in every platform project.

hartez replied:
Okay, moved to internal helper classes on each platform.

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,142 @@
+		public static bool IsOneOf(this PropertyChangedEventArgs args, params BindableProperty[] properties)
+		{
+			CoreFlags.VerifyCollectionViewFlagEnabled();
+			for (int n = 0; n < properties.Length; n++)
```
properties is an array, `foreach` is free

hartez replied:
Fixed.

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	[RenderWith(typeof(_CarouselViewRenderer))]
+	public class CarouselView : ItemsView
```
move to it's own file

hartez replied:
Done.

StephaneDelcroix commented on Xamarin.Forms.Core/Items/ItemsView.cs:
```diff
@@ -0,0 +1,142 @@
+			CoreFlags.VerifyCollectionViewFlagEnabled(constructorHint: nameof(ItemsView));
+		}
+
+		// TODO hartez 2018/06/24 11:37:00 Give DisplayMemberPath some thought	
```
not now, drop the //TODO

hartez commented:
@StephaneDelcroix 

> there are still //TODO left

That's because there are still things left to do. It's a work-in-progress, that's why it's behind the flag. 
Or do you just want the TODO comments themselves removed?

> we don't want ToString()

Why don't we want to implement ToString()?

hartez commented on Xamarin.Forms.ControlGallery.Android/FormsAppCompatActivity.cs:
```diff
@@ -41,9 +41,12 @@ protected override void OnCreate(Bundle bundle)
-			Forms.SetFlags("Fake_Flag"); // So we can test for flag initialization issues
+			// Fake_Flag is here so we can test for flag initialization issues
+			// CollectionView lets us test CollectionView stuff until it's officially released
+			Forms.SetFlags("Fake_Flag", "CollectionView_Experimental"); 
```
Done.

hartez commented on Xamarin.Forms.Core/Application.cs:
```diff
@@ -7,6 +7,7 @@
 using Xamarin.Forms.Internals;
 using Xamarin.Forms.Platform;
 using System.Diagnostics;
+using System.Linq;
```
Removed.

hartez commented on Xamarin.Forms.Platform.UAP/ListViewRenderer.cs:
```diff
@@ -17,6 +17,7 @@
 using WApp = Windows.UI.Xaml.Application;
 using Xamarin.Forms.Internals;
 using Xamarin.Forms.PlatformConfiguration.WindowsSpecific;
+using WItemsControl = Windows.UI.Xaml.Controls.ItemsControl;
```
Removed.

PureWeen commented:
@hartez   rebase to fix conflicts then let's run UI tests again with flags off

hartez commented on Xamarin.Forms.Core/Items/GridItemsLayout.cs:
```diff
@@ -0,0 +1,28 @@
+			set => SetValue(SpanProperty, value);
+		}
+
+		public GridItemsLayout([Parameter("Span")] int span, [Parameter("Orientation")] ItemsLayoutOrientation orientation) :
```
My bad, that's left over from before we decided to allow spans to be changed. Fixed now.

hartez commented on Xamarin.Forms.Core/Items/ListItemsLayout.cs:
```diff
@@ -0,0 +1,21 @@
+
+		// TODO hartez 2018/08/29 20:31:54 Need something like these previous two, but as a carousel default	
+
+		public override string ToString()
```
Dropped.

hartez commented on Xamarin.Forms.Core/Items/GridItemsLayout.cs:
```diff
@@ -0,0 +1,28 @@
+			Span = span;
+		}
+
+		public override string ToString()
```
Dropped.

hartez commented:
@PureWeen 
> Can we get a public API diff caused by this PR? 

Here are all the new public classes/interfaces:

Core, Added:

- public interface IItemsLayout
- public enum SnapPointsType
- public enum SnapPointsAlignment
- public enum ItemsLayoutOrientation
- public class ScrollToRequestEventArgs
- public abstract class ItemsLayout
- public class ListItemsLayout
- public class GridItemsLayout 
- public class ItemsView
- public class CollectionView
- public class CarouselView

UWP, Added:

- public class ItemContentControl
- public class CollectionViewRenderer

iOS, Added:

- public abstract class ItemsViewCell 
- public abstract class DefaultCell 
- public abstract class TemplatedCell 
- public abstract class ItemsViewLayout 
- public class CarouselViewRenderer
- public class CollectionViewRenderer 

Android, Added:

- public class SnapManager 
- public class PositionalSmoothScroller 
- public class ItemsViewRenderer 
- public class ItemsViewAdapter 	
- public class EmptyViewAdapter 
- public class CollectionViewRenderer 
- public class CarouselViewRenderer
